### ISSUE TITLE
Settings randomly revert back

### ISSUE THREAD

**Issue by freemo** [bitnami/bitnami-docker-discourse#109]
**Description**

I have a discourse instance that has been up for a week now and it has some of the weirdest symptoms that I cant solve partly because there is nothing at all showing up in the logs that are useful.

The most noticeable symptom is that if i go into the settings and change a setting, then hit the green check mark it accepts it and all looks good. If i then refresh the settings page there is a good chance it might show the old setting or the new. If i continue to refresh sometimes it shows one setting, sometimes the other, no matter how many times i refresh it may switch back and forth multiple times.

Worse yet it isnt just in the settings screen itself but sometimes the functionality of the server itself sees to change. For example if i try uploading a new logo it may sometimes show the old and sometimes the new.

To answer the obvious question: no I am not running multiple instances behind a load balancer, single instance.

I’m not sure if it is related but i also notice the discourse math plugin doesnt render right either. While the equations do show correctly int he preview once i post it the javascript doesnt appear to get included and thus doesnt render properly. The math issue isnt the priority here so just mentioning it here in case it offers any more clues.

I’m really stuck here, with the logs all looking normal I’ not even sure what to post to get help. Any ideas would be great.

**Steps to reproduce the issue:**

Seems to happen everytime i setup an environment on ECS. I cant narrow down the cause

**Describe the results you received:**

If i go into the settings and change a setting, then hit the green check mark it accepts it and all looks good. If i then refresh the settings page there is a good chance it might show the old setting or the new. If i continue to refresh sometimes it shows one setting, sometimes the other, no matter how many times i refresh it may switch back and forth multiple times.

**Describe the results you expected:**

I expect settings to be retained

**Additional information you deem important (e.g. issue happens only occasionally):**

I checked the database and the value is being set properly in site_settings table

**Version**

- Output of `docker version`:

```
Docker version 18.06.1-ce, build e68fc7a215d7133c34aa18e3b72b4a21fd0c6136
```

- Output of `docker info`:

```
Containers: 15
 Running: 15
 Paused: 0
 Stopped: 0
Images: 12
Server Version: 18.06.1-ce
Storage Driver: devicemapper
 Pool Name: docker-docker--pool
 Pool Blocksize: 524.3kB
 Base Device Size: 10.74GB
 Backing Filesystem: ext4
 Udev Sync Supported: true
 Data Space Used: 11.87GB
 Data Space Total: 23.33GB
 Data Space Available: 11.46GB
 Metadata Space Used: 6.509MB
 Metadata Space Total: 25.17MB
 Metadata Space Available: 18.66MB
 Thin Pool Minimum Free Space: 2.333GB
 Deferred Removal Enabled: true
 Deferred Deletion Enabled: true
 Deferred Deleted Device Count: 0
 Library Version: 1.02.135-RHEL7 (2016-11-16)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 468a545b9edcd5932818eb9de8e72413e616e86e
runc version: 69663f0bd4b60df09991c08812a60108003fa340
init version: fec3683
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.14.133-88.112.amzn1.x86_64
Operating System: Amazon Linux AMI 2018.03
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 31.05GiB
Name: ip-10-0-0-127
ID: YM6O:MFCA:FULC:73KT:MZTX:O6TM:I63D:T3LI:HDCO:ZU2K:4VMD:VAKX
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, Docker for MAC, physical, etc.):**

Running in AWS ECS

**Comment by gregoryboue** [comment]
Hi,

You can see : https://github.com/bitnami/bitnami-docker-discourse/issues/107 
I think it's the same problem, to make settings stable i find a temporary solution, go to your /opt/bitnami/discourse folder into the container and launch RAILS_ENV=production bundle exec rake restart

**Comment by marcosbc** [comment]
Hi, just wanted to let you know we're investigating this issue.

It could be related to some caching mechanism but right now we have no idea. We'll get back if we have any update.

In the meantime, have you been able to do this in a reproducible way? E.g. starting the container, changing X setting and reloading 3 times, and then it's back to original. Thanks!

**Comment by freemo** [comment]
@marcosbc Hi thanks. 

The issue always occurs for me. I am not sure what exactly it takes to reproduce it i just know ive setup the server more than once and reset it countless times and the issue is always there.

@gregoryboue I will test to see if that helps, thanks.

**Comment by freemo** [comment]
@gregoryboue I had a chance to test your work around, it didnt help.

The problem persists.

**Comment by freemo** [comment]
Incase it is useful here is the container configuration JSON I use to bring up the services (this is on AWS ECS):

```
[{
    "name": "qoto-discourse",
    "image": "bitnami/discourse:2",
    "portMappings": [{
        "containerPort": 3000,
        "hostPort": 3000
    }],
    "memory": 2048,
    "mountPoints": [{
        "sourceVolume": "discourse-data",
        "containerPath": "/bitnami",
        "readOnly": false
    },{
        "sourceVolume": "nginx-logs",
        "containerPath": "/var/log/nginx",
        "readOnly": false
    }],
    "environment": [{
        "name": "POSTGRESQL_HOST",
        "value": "redacted.rds.amazonaws.com"
    },{
        "name": "POSTGRESQL_ROOT_USER",
        "value": "redacted"
    },{
        "name": "POSTGRESQL_ROOT_PASSWORD",
        "value": "redacted"
    },{
        "name": "DISCOURSE_POSTGRESQL_NAME",
        "value": "redacted"
    },{
        "name": "DISCOURSE_USERNAME",
        "value": "redacted"
    },{
        "name": "DISCOURSE_EMAIL",
        "value": "freemo@gmail.com"
    },{
        "name": "DISCOURSE_PASSWORD",
        "value": "redacted"
    },{
        "name": "DISCOURSE_POSTGRESQL_USERNAME",
        "value": "redacted"
    },{
        "name": "DISCOURSE_POSTGRESQL_PASSWORD",
        "value": "redacted"
    },{
        "name": "DISCOURSE_HOSTNAME",
        "value": "discourse.qoto.org"
    },{
        "name": "REDIS_HOST",
        "value": "redacted.0001.use1.cache.amazonaws.com"
    },{
        "name": "SMTP_HOST",
        "value": "redacted.amazonaws.com"
    },{
        "name": "SMTP_PORT",
        "value": "587"
    },{
        "name": "SMTP_USER",
        "value": "redacted"
    },{
        "name": "SMTP_PASSWORD",
        "value": "redacted"
    },{
        "name": "DISCOURSE_SKIP_INSTALL",
        "value": "no"
    },{
        "name": "DISCOURSE_SITENAME",
        "value": "QOTO Discourse"
    },{
        "name": "VIRTUAL_HOST",
        "value": "discourse.qoto.org"
    },{
        "name": "LETSENCRYPT_HOST",
        "value": "discourse.qoto.org"
    },{
        "name": "VIRTUAL_PORT",
        "value": "3000"
    },{
        "name": "LETSENCRYPT_EMAIL",
        "value": "jeffrey.freeman@qoto.org"
    },{
        "name": "DEBUG",
        "value": "true"
    }]
}, {
    "name": "qoto-discourse-sidekiq",
    "image": "bitnami/discourse:2",
    "memory": 2048,
    "mountPoints": [{
        "sourceVolume": "sidekiq-data",
        "containerPath": "/bitnami",
        "readOnly": false
    },{
        "sourceVolume": "nginx-logs",
        "containerPath": "/var/log/nginx",
        "readOnly": false
    }],
    "environment": [{
        "name": "POSTGRESQL_HOST",
        "value": "redacted.rds.amazonaws.com"
    },{
        "name": "DISCOURSE_POSTGRESQL_NAME",
        "value": "redacted"
    },{
        "name": "DISCOURSE_POSTGRESQL_USERNAME",
        "value": "redacted"
    },{
        "name": "DISCOURSE_POSTGRESQL_PASSWORD",
        "value": "redacted"
    },{
        "name": "DISCOURSE_HOSTNAME",
        "value": "discourse.qoto.org"
    },{
        "name": "REDIS_HOST",
        "value": "redacted.ng.0001.use1.cache.amazonaws.com"
    },{
        "name": "SMTP_HOST",
        "value": "redacted.amazonaws.com"
    },{
        "name": "SMTP_PORT",
        "value": "587"
    },{
        "name": "SMTP_USER",
        "value": "redacted"
    },{
        "name": "SMTP_PASSWORD",
        "value": "redacted"
    },{
        "name": "DISCOURSE_HOST",
        "value": "discourse.qoto.org"
    },{
        "name": "DISCOURSE_PORT",
        "value": "3000"
    },{
        "name": "DEBUG",
        "value": "true"
    }],
    "command": [
            "nami",
            "start",
            "--foreground",
            "discourse-sidekiq"
    ]
}]

```

**Comment by gregoryboue** [comment]
Hi,

To be more precise about the workaround, it doesn't fix the problem for me, it's just that when i modify settings on GUI panel and launch the command into the container, then the behavior is ok.

But if i modify another settings without launch the command i still have the problem.

**Comment by marcosbc** [comment]
Hi, we have been able to reproduce this issue. We're investigating this issue, in case it's related with Discourse or with the Bitnami configuration.

Unfortunately we cannot provide an ETA for when we expect this to be fixed, but we'll keep this issue updated with any progress.

**Comment by gregoryboue** [comment]
@marcosbc thanks you, i will wait :-)

**Comment by freemo** [comment]
@marcosbc I eagerly await to hear the progress. No one on standard discourse appears to have this problem (I spoke with them). It is most likely bitnami specific.

**Comment by marcosbc** [comment]
Hi, just wanted to update this issue as we have spent some time checking this.

TL;DR; Passenger is not working properly with Discourse. If you're curious to know how we found the issue, we have included the investigation details below. Since this is not an officially supported configuration, it's unlikely this will ever get fixed on Discourse's side.

We are going to open an internal discussion on improvements we can make to the Bitnami Discourse container. Keep in mind these are pretty big as it is likely we may need to change Passenger by another server such as Unicorn or Puma, so it may take some time before these changes are implemented.

In any case, we'll keep this issue updated with any progress. Please let us know if you have any questions.

______

After increasing the PostgreSQL verbosity to [show all SQL commands](https://stackoverflow.com/questions/722221/how-to-log-postgresql-queries) we could see the `site_settings` table being only being fetched after reboots, meaning posterior changes should be handled by ActiveDirectory, and implying there is some kind of cache mechanism going on.

It is easy to verify that the problem is code caching, as the value in the PostgreSQL database is always correct, but still the code shows the wrong value: The HTTP request to the Discourse API sometimes show the outdated values, and debugging the code (specifically right after `lib/site_settings.rb` line 233) you can see the wrong values are retrieved, indicating the issue.

We've identified two workarounds for this issue: a) Restart Apache OR b) Remove Passenger cache in `/tmp/passenger.*/apps.s`.

We thought of the following things that did not help workaround this:

- Clearing Rails cache (`ActiveRecord::Base.connection.query_cache.clear`): https://stackoverflow.com/questions/3740705/clearing-activerecord-cache
- Running `rake tmp:cache:clear`

We confirmed this does NOT happen in the official Discourse Docker installation, so it seemed to be some kind in one of the Ruby components or configurations bundled in the image. The main difference between Bitnami's image and the official Docker image is we use Passenger for running Discourse, the official one uses Unicorn.

With that we tested the following scenarios:

- Official installation with Unicorn: Not affected.
- Official installation with Passenger: **Affected**.
- Bitnami installation with Unicorn: Not affected.
- Bitnami installation with Passenger: **Affected**.
- (Bitnami installation with Puma: Not affected.)

With that it is easy to confirm the issue being Passenger, but we could not find any setting for configuring the cache, especially for ActiveRecord requests.

**Comment by marcosbc** [comment]
Hi, we have been able to identify a configuration fix in Passenger to avoid this issue, and also avoiding the need to change to a different server. The release version for the container should be `2.3.2-debian-9-r48` and `2.3.2-ol-7-r47`, and should be out later today or tomorrow morning.

We're adding an entry in Notable Changes describing the details. You should be able to update the container without requiring changes on your side.

**Comment by freemo** [comment]
@marcosbc Thanks so much for all the hard work. You all rock!

**Comment by marcosbc** [comment]
Hi @freemo, just wanted to let you know Discourse was released with the mentioned fix.

Find the details in [Notable changes](https://github.com/bitnami/bitnami-docker-discourse#232-debian-9-r48-and-232-ol-7-r47), and the documentation for `DISCOURSE_PASSENGER_SPAWN_METHOD` in [Environment variables](https://github.com/bitnami/bitnami-docker-discourse#environment-variables).
